app/components-react/shared/SearchablePages.tsx

```diff
@@ -0,0 +1,144 @@
+import React, { useState, useRef, useEffect } from 'react';
+import ReactDOM from 'react-dom';
+import Spinner from './Spinner';
+import Mark from 'mark.js';
+import styles from './SearchablePages.m.less';
+import { TCategoryName } from 'services/settings';
+import { SETTINGS_CONFIG } from 'components-react/windows/Settings';
+import { createRoot } from 'components-react/root/ReactRoot';
+
+interface ISearchablePagesProps {
+  page: TCategoryName;
+  pages: TCategoryName[];
+  searchStr: string;
+  onSearchCompleted?: (results: string[]) => void;
+}
+
+export default function SearchablePages(p: React.PropsWithChildren<ISearchablePagesProps>) {
+  const pagesInfo = useRef<PartialRec<TCategoryName, string> | null>(null);
+  const pageRef = useRef<HTMLDivElement>(null);
+  const [loading, setLoading] = useState(false);
+
+  // Search pages when receiving a search string input
+  useEffect(() => {
+    if (loading || p.searchStr === '' || p.searchStr.length < 3) return;
+    async function searchHandler() {
+      // build cache of info on pages if it doesn't exist
+      if (!pagesInfo.current) await scanPages();
+
+      // find matching info in cache
+      const searchResultPages: TCategoryName[] = [];
+      if (!pagesInfo.current) return;
+      for (const pageName of Object.keys(pagesInfo.current)) {
+        const pageText = pagesInfo.current[pageName as TCategoryName];
+        if (pageText && pageText.match(new RegExp(p.searchStr, 'ig'))) {
+          searchResultPages.push(pageName as TCategoryName);
+        }
+      }
+      if (searchResultPages.length === 0) return;
+      // we send out search results to receive as props here
+      // in order to properly display the searched page
+      p.onSearchCompleted && p.onSearchCompleted(searchResultPages);
+    }
+
+    searchHandler().then(highlightPage);
+  }, [p.searchStr, loading]);
+
+  // Highlight new page if necessary
+  useEffect(() => {
+    if (loading || p.searchStr === '' || p.searchStr.length < 3) return;
+    highlightPage();
+  }, [p.page]);
+
+  // /**
+  //  * fetch and cache all text information from settings pages
+  //  */
+  async function scanPages() {
+    setLoading(true);
+    pagesInfo.current = {};
+
+    for (const page of p.pages) {
+      const component = SETTINGS_CONFIG[page].component;
+
+      // collect the page text and text from inputs
+      pagesInfo.current[page] = await grabReactTextContent(component);
+    }
+
+    setLoading(false);
+  }
+
+  async function grabReactTextContent(component: React.FunctionComponent<any>): Promise<string> {
```

Comment:
same as `createRoot1, if it doesn't create problems, I think it can be made generic to not lose the type info.